subsys/net/lib/coap/coap.c

```diff
@@ -454,6 +454,11 @@ static int parse_option(uint8_t *data, uint16_t offset, uint16_t *pos,
 	uint8_t opt;
 	int r;
 
+	/* Check if there are options to parse */
+	if (offset == max_len) {
+		return 0;
+	}
+
 	r = read_u8(data, offset, pos, max_len, &opt);
 	if (r < 0) {
 		return r;
```

Comment:
Actually this `read_u8()` should return `r == -EINVAL` when `offset == max_len`, in which case you can do `return 0;` instead of `return r;`. 

**But**, I don't think `parse_option()` should return 0 in case of **no parsed** message. Simply you no longer can distinguish between:
1) `parse_option()` that parsed option successfully and reached end of buffer (max_len),
2) `parse_option()` did not parse any option, because we were already at the end of buffer (offset == max_len).

So I think it is `coap_packet_parse()` and/or `coap_packet_parse()` that should be "fixed".

Reply:
I agree, `parse_option()` shouldn't be called when there is no payload to be parsed.
Upon further investigation of `coap_packet_parse()` and `coap_find_options()`, it seems that [`coap_packet_parse()` is actually already checking for this](https://github.com/zephyrproject-rtos/zephyr/blob/cfe0849072c81c1314f733babcc9a87340c5d595/subsys/net/lib/coap/coap.c#L592). `coap_find_options()`, does not. Please see my new commit.

Reply:
To be honest I already see some ambiguity in the `parse_option()` function and how `coap_find_options()` uses it. Correct me if I'm wrong (I've not run any tests, just analyzed the code) but consider two scenarios:
1. The packet being parsed have a set of options and no payload - an option is parsed on each iteration, on the final call the last option is parsed and `0` is returned.
2.  The packet being parsed have a set of options and a payload - an option is parsed on each iteration but the last one, on the final call the payload marker is detected and 0 returned, no option was parsed. If you look at the [`coap_find_options()`](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/coap/coap.c#L643) implementation, the option value is being checked anyway, which would hold some rubbish from the function perspective (well, whatever the `options` struct was initialized with).

I guess it's not something to be addressed in this PR (the change in `coap_find_options` looks fine for the issue reported) but we should likely leave an issue for this (if I'm correct of course).

Reply:
> I guess it's not something to be addressed in this PR (the change in `coap_find_options` looks fine for the issue reported) but we should likely leave an issue for this (if I'm correct of course).

Doesn't the added check in this PR solve that second scenario? Unfortunately I can't budget more time for this at this moment. But I think the PR should indeed be fine for now.

Reply:
I don't think so, the change in the PR fixes the case when there is no option/payload at all. The second scenario is about having both, options and payload in the packet. But yes, the PR looks good to me with regards to the issue it addresses.